libpromises/evalfunction.c

```diff
@@ -2359,12 +2362,26 @@ static FnCallResult FnCallExecResult(ARG_UNUSED EvalContext *ctx, ARG_UNUSED con
         }
     }
 
-    if (GetExecOutput(command, &buffer, &buffer_size, shelltype, output_select, NULL))
+    int exit_code;
+
+    if (GetExecOutput(command, &buffer, &buffer_size, shelltype, output_select, &exit_code))
     {
         Log(LOG_LEVEL_VERBOSE, "%s ran '%s' successfully", fp->name, command);
-        FnCallResult res = FnReturn(buffer);
-        free(buffer);
-        return res;
+        if (StringEqual(function, "execresult"))
+        {
+            FnCallResult res = FnReturn(buffer);
+            free(buffer);
+            return res;
+        }
+        else
+        {
+            assert(StringEqual(function, "execresult_as_data"));
+            JsonElement *result = JsonObjectCreate(2);
+            JsonObjectAppendInteger(result, "exit_code", exit_code);
+            JsonObjectAppendString(result, "output", buffer);
+            free(buffer);
```

Comment:
Do we have a function to append a string without creating a copy?

Reply:
Nope